File: code_snippets/go/guides/retries.go
package guides

import (
	restate "github.com/restatedev/sdk-go"
	"log/slog"
	"time"
)

type MyService struct{}

func (MyService) Greet2(ctx restate.Context, name string) error {
	// <start_here>
	result, err := restate.Run(ctx,
		func(ctx restate.RunContext) (string, error) {
			return writeToOtherSystem()
		},
		// <mark_1>
		// After 10 seconds, give up retrying
		restate.WithMaxRetryDuration(time.Second*10),
		// On the first retry, wait 100 milliseconds before next attempt
		restate.WithInitialRetryInterval(time.Millisecond*100),
		// Grow retry interval with factor 2
		restate.WithRetryIntervalFactor(2.0),
		// </mark_1>
	)
	if err != nil {
		return err
	}
	// <end_here>

	_ = result

	return nil
}

func (MyService) Greet(ctx restate.Context, name string) error {
	// <start_catch>
	result, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return writeToOtherSystem()
	})
	if err != nil {
		if restate.IsTerminalError(err) {
			// Handle the terminal error: undo previous actions and
			// propagate the error back to the caller
		}
		return err
	}
	// <end_catch>

	_ = result

	return nil
}

// <start_raw>
func (MyService) myHandler(ctx restate.Context) (string, error) {
	// !mark
	rawRequest := ctx.Request().Body
	decodedRequest, err := decodeRequest(rawRequest)
	if err != nil {
		if restate.IsTerminalError(err) {
			// Propagate to DLQ/catch-all handler
		}
		return "", err
	}

	// ... rest of your business logic ...
	return decodedRequest, nil
}

// <end_raw>

func (MyService) myTimeoutHandler(ctx restate.Context) error {
	// <start_timeout>
	awakeable := restate.Awakeable[string](ctx)
	// !mark
	timeout := restate.After(ctx, 5*time.Second)
	// !mark
	selector := restate.Select(ctx, awakeable, timeout)
	switch selector.Select() {
	case awakeable:
		result, err := awakeable.Result()
		if err != nil {
			return err
		}
		slog.Info("Awakeable resolved first with: " + result)
	case timeout:
		if err := timeout.Done(); err != nil {
			return err
		}
		slog.Info("Timeout hit first")
	}
	// <end_timeout>
	return nil
}

func decodeRequest(request []byte) (string, error) {
	return "req", nil
}

func writeToOtherSystem() (string, error) {
	return "", nil
}

File: code_snippets/rust/src/guides/retries.rs
#![allow(unused_variables)]
#![allow(dead_code)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::question_mark)]

use restate_sdk::prelude::*;
use std::time::Duration;

#[restate_sdk::service]
trait MyService {
    async fn handle() -> Result<(), TerminalError>;
    async fn my_handler(request: Vec<u8>) -> Result<(), HandlerError>;
}

struct MyServiceImpl;
impl MyService for MyServiceImpl {
    async fn handle(&self, ctx: Context<'_>) -> Result<(), TerminalError> {
        // <start_here>
        // <mark_1>
        let my_run_retry_policy = RunRetryPolicy::default()
            .initial_delay(Duration::from_millis(100))
            .exponentiation_factor(2.0)
            .max_delay(Duration::from_millis(1000))
            .max_attempts(10)
            .max_duration(Duration::from_secs(10));
        // </mark_1>
        ctx.run(|| write_to_other_system())
            .retry_policy(my_run_retry_policy)
            .await?;
        // <end_here>

        // <start_catch>
        // Fails with a terminal error after 3 attempts or if the function throws one
        if let Err(e) = ctx
            .run(|| write_to_other_system())
            .retry_policy(RunRetryPolicy::default().max_attempts(3))
            .await
        {
            // Handle the terminal error: undo previous actions and
            // propagate the error back to the caller
            return Err(e);
        }
        // <end_catch>

        // <start_terminal_error>
        Err(TerminalError::new("This is a terminal error"))
        // <end_terminal_error>
    }

    // <start_raw>
    // Use Vec<u8> to represent a binary request
    // !mark[/request: Vec<u8>/] blue
    async fn my_handler(&self, ctx: Context<'_>, request: Vec<u8>) -> Result<(), HandlerError> {
        let decoded_request = decode_request(&request)?;

        // ... rest of you business logic ...

        Ok(())
    }
    // <end_raw>
}

fn decode_request(p0: &[u8]) -> Result<String, HandlerError> {
    Ok(String::from("Hello"))
}

async fn write_to_other_system() -> Result<String, HandlerError> {
    Ok("Hello".to_string())
}
